File: Big0/Javascript/big_0.js
//    0(1)  - Constant Time
function getFirstElement(arr){
    return arr[0]; // Always 1 step regardless of array size
}

//0(n) - Linear Time
function printAllItems(arr){
    arr.forEach(element => {
        console.log(element);
    });// One print per element
}

// 0(n^2) - Quadratic Time
function printAllPairs(arr){
    for(let i = 0; i < arr.length; i++){
        for(let j = 0; j < arr.length; j ++){
            console.log(arr[i], arr[j]);
        }
    }
}

//0(log n) - Logarithmic Time (Binary search)
function binarySearch(arr, target){
    let left = 0;
    let right = arr.length - 1;
    while(left <= right){
        const mid = Math.floor((left + right) / 2);
        if(arr[mid] === target) return mid;
        if(arr[mid] < target) left = mid + 1;
        else right = mid -1;
    }
    return -1
}

File: Big0/Python/big_0.py
# 0(1) - constant Time
def get_first_element(arr):
    return arr[0]

# 0(n) - linear Time
def print_all_items(arr):
    for item in arr:
        print(item)

# 0(n^2) - Quadratic Time
def print_all_pairs(arr):
    for i in arr:
        for j in arr:
            print(i, j)

# 0(log n) - Logarithmic Time
def binary_search(arr, target):
    left = 0
    right = len(arr) - 1
    while left <= right:
        mid = (left + right) / 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid -1
    return -1